analyzer/codechecker_analyzer/analyzers/clangsa/ctu_autodetection.py

```diff
@@ -0,0 +1,134 @@
+
+from codechecker_common.logger import get_logger
+
+LOG = get_logger('analyzer.clangsa')
+
+
+def test_binary_existence(tool_path):
+    """ Test the executable for existence by checking its version. """
+    try:
+        version = subprocess.check_output([tool_path, '-version'])
+    except (subprocess.CalledProcessError, OSError):
+        version = 'ERROR'
+    return version != 'ERROR'
+
+
+class CTUAutodetection(object):
+
+    def __init__(self, analyzer_binary):
+        self.__analyzer_binary = analyzer_binary
+        self.__analyzer_version_info = None
+
+    @property
+    def analyzer_version_info(self):
+        """ Returns the relevant parameters of the analyzer by parsing the
+        output of the analyzer binary when called with version flag. """
+
+        if self.__analyzer_binary is None:
+            return False
+
+        if self.__analyzer_version_info is not None:
+            return self.__analyzer_version_info
+
+        try:
+            analyzer_output = subprocess.check_output(
+                [self.__analyzer_binary, '--version'])
+        except (subprocess.CalledProcessError, OSError):
+            return False
+
+        version_pattern = re.compile(
+            r'clang version (?P<major_version>[0-9]+)'
+            r'\.(?P<minor_version>[0-9]+)\.(?P<patch_version>[0-9]+)\s+'
+            r'\((?P<clang_repository>[^\s]*)\s+(?P<clang_commit>[^\s]*)\)\s+'
+            r'\((?P<llvm_repository>[^\s]*)\s+(?P<llvm_commit>[^\s]*)\)\s+'
+            r'Target: (?P<target>[^\s]*)\s+'
+            r'Thread model: (?P<thread_model>[^\s]*)\s+'
+            r'InstalledDir: (?P<installed_dir>[^\s]*).*')
+
+        match = (version_pattern.search(analyzer_output))
+
+        if not match:
+            return False
+
+        self.__analyzer_version_info = match.groupdict()
+        return self.__analyzer_version_info
+
+    @property
+    def major_version(self):
+        """ Returns the major version of the analyzer, which is used for
+        CTU analysis. """
+
+        if not self.analyzer_version_info:
+            return False
+
+        try:
+            return (int)(self.analyzer_version_info['major_version'])
+        except ValueError:
+            return False
+
+    @property
+    def installed_dir(self):
+        """ Returns the installed directory of the analyzer, which is used for
+        CTU analysis. """
+
+        if not self.analyzer_version_info:
+            return False
+
+        return self.analyzer_version_info['installed_dir']
+
+    @property
+    def mapping_tool_path(self):
+        """ Returns the path of the mapping tool, which is assumed to be
+        located inside the installed directory of the analyzer."""
+
+        if not self.analyzer_version_info:
+            return False
+
+        major_version = self.analyzer_version_info['major_version']
+        installed_dir = self.analyzer_version_info['installed_dir']
+
+        tool_name = 'clang-extdef-mapping' if major_version > 7 else\
```

Comment:
Wouldn't it be better to check the existence of the `clang-extdef-mapping` tool? If it doesn't exist than use the `clang-func-mapping` tool.

Reply:
This wouldn't be a bad idea, however the specific Clang requires its specific tool, and also it could be postfixed by the major version number.